### PR TITLE
add new BaseClientImpl that uses SimpleHttpClient

### PR DESCRIPTION
**Pull request by kimyen**

### CODE REVIEW

File: client/synapseJavaClient/src/main/java/org/sagebionetworks/client/ClientUtils.java [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+		Header contentTypeHeader = response.getFirstHeader(HttpHeaders.CONTENT_TYPE);
+		return getCharacterSetFromContentTypeHeader(contentTypeHeader);
+	}
+
+	/**
+	 * 
+	 * @param contentTypeHeader
+	 * @return
+	 */
+	public static Charset getCharacterSetFromContentTypeHeader(Header contentTypeHeader) {
```
Comment (john-hill) [inline client/synapseJavaClient/src/main/java/org/sagebionetworks/client/ClientUtils.java:164]: if this signature is not used then combine with getCharacterSetFromResponse()

File: client/synapseJavaClient/src/test/java/org/sagebionetworks/client/ClientUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+		assertFalse(ClientUtils.is200sStatusCode(300));
+	}
+
+	@Test (expected = SynapseUnauthorizedException.class)
+	public void testThrowUnauthorizedException() throws Exception{
+		JSONObject response = new JSONObject();
+		response.append(ERROR_REASON_TAG, "test");
+		ClientUtils.throwException(401, response);
+	}
+
```
Comment (john-hill) [inline client/synapseJavaClient/src/test/java/org/sagebionetworks/client/ClientUtilsTest.java:69]: need one test that looks at the message of the exception.

File: client/synapseJavaClient/src/test/java/org/sagebionetworks/client/ClientUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+		assertEquals("some reason", actual.get("reason"));
+	}
+
+
+	@Test (expected = SynapseBadRequestException.class)
+	public void testConvertResponseBodyToJSONAndThrowExceptionFor300() throws Exception{
+		when(mockResponse.getContent()).thenReturn("{\"reason\":\"some reason\"}");
+		when(mockResponse.getStatusCode()).thenReturn(400);
+		ClientUtils.convertResponseBodyToJSONAndThrowException(mockResponse);
+	}
```
Comment (john-hill) [inline client/synapseJavaClient/src/test/java/org/sagebionetworks/client/ClientUtilsTest.java:154]: check message here as well.

File: client/synapseJavaClient/src/main/java/org/sagebionetworks/client/ClientUtils.java [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+				requestUrl = new URL(endpoint + uri);
+			}
+			builder = new URIBuilder(requestUrl.toURI());
+			if (parameters != null) {
+				for (Map.Entry<String,String> entry : parameters.entrySet()) {
+					builder.addParameter(entry.getKey(), entry.getValue());
+				}
+			} else {
+				// TODO: do we need this?
+				builder.addParameter(AuthorizationConstants.DOMAIN_PARAM, SYNAPSE_DOMAIN.name());
```
Comment (john-hill) [inline client/synapseJavaClient/src/main/java/org/sagebionetworks/client/ClientUtils.java:201]: Need to remove this

File: client/synapseJavaClient/src/main/java/org/sagebionetworks/client/BaseClientImpl.java [hunk 5]
```diff
@@ -1,74 +1,873 @@
...
+	private static final String PUT_METHOD = "PUT";
+	private static final String DELETE_METHOD = "DELETE";
+
+	private SimpleHttpClient simpleHttpClient;
+
+	private String userAgent;
+	private String authEndpoint;
+	private Map<String, String> defaultGETDELETEHeaders;
+	private Map<String, String> defaultPOSTPUTHeaders;
+	private JSONObject profileData;
```
Comment (john-hill) [inline client/synapseJavaClient/src/main/java/org/sagebionetworks/client/BaseClientImpl.java:73]: profileData should not be used anymore.

File: client/synapseJavaClient/src/main/java/org/sagebionetworks/client/BaseClientImpl.java [hunk 6]
```diff
@@ -1,74 +1,873 @@
...
+		this(userAgent, null);
+	}
+
+	public BaseClientImpl(String userAgent, SimpleHttpClientConfig config) {
+		this.userAgent = userAgent;
+		this.simpleHttpClient = new SimpleHttpClientImpl(config);
+
+		this.authEndpoint = DEFAULT_AUTH_ENDPOINT;
+		this.defaultGETDELETEHeaders = new HashMap<String, String>();
+		this.defaultGETDELETEHeaders.put("Accept", "application/json; charset="+SYNAPSE_ENCODING_CHARSET);
```
Comment (john-hill) [inline client/synapseJavaClient/src/main/java/org/sagebionetworks/client/BaseClientImpl.java:88]: constant

File: client/synapseJavaClient/src/main/java/org/sagebionetworks/client/BaseClientImpl.java [hunk 7]
```diff
@@ -1,74 +1,873 @@
...
+	 * @category Upload & Download
+	 * @param endpoint
+	 * @param uri
+	 * @return
+	 * @throws ClientProtocolException
+	 * @throws IOException
+	 * @throws FileNotFoundException
+	 * @throws SynapseException 
+	 */
+	protected String downloadZippedFileToString(String endpoint, String uri)
```
Comment (john-hill) [inline client/synapseJavaClient/src/main/java/org/sagebionetworks/client/BaseClientImpl.java:209]: this is only used for wiki.  should be deprecated.

File: client/synapseJavaClient/src/main/java/org/sagebionetworks/client/BaseClientImpl.java [hunk 8]
```diff
@@ -1,74 +1,873 @@
...
+	/**
+	 * Asymmetrical put where a response JSONEntity is not expected.
+	 * 
+	 * @category JSONEntity Requests
+	 * @param endpoint
+	 * @param url
+	 * @param requestBody
+	 * @throws SynapseException
+	 */
+	protected void asymmetricalPut(String endpoint, String url, JSONEntity requestBody) throws SynapseException {
```
Comment (john-hill) [inline client/synapseJavaClient/src/main/java/org/sagebionetworks/client/BaseClientImpl.java:468]: void put

File: client/synapseJavaClient/src/main/java/org/sagebionetworks/client/BaseClientImpl.java [hunk 9]
```diff
@@ -1,74 +1,873 @@
...
+	 * @category JSONEntity Requests
+	 * @param endpoint
+	 * @param url
+	 * @param requestBody
+	 * @param returnClass
+	 * @param params
+	 * @param errorHandler
+	 * @throws SynapseException
+	 */
+	protected void asymmetricalPost(String endpoint, String url, JSONEntity requestBody,
```
Comment (john-hill) [inline client/synapseJavaClient/src/main/java/org/sagebionetworks/client/BaseClientImpl.java:516]: void post

File: client/synapseJavaClient/src/main/java/org/sagebionetworks/client/BaseClientImpl.java [hunk 10]
```diff
@@ -1,74 +1,873 @@
...
+	}
+
+	/**
+	 * @param requestUrl
+	 * @param requestMethod
+	 * @param requestContent
+	 * @param requestHeaders
+	 * @return
+	 */
+	protected SimpleHttpResponse performRequestWithRetry(final String requestUrl,
```
Comment (john-hill) [inline client/synapseJavaClient/src/main/java/org/sagebionetworks/client/BaseClientImpl.java:724]: check if the portal was using retryRequestIfServiceUnavailable=false

Comment (kimyen) [reply client/synapseJavaClient/src/main/java/org/sagebionetworks/client/BaseClientImpl.java:724]: portal wasn't using this boolean

**Pull request merged**